### PR TITLE
Adding possibility to select the compression type for the OpenEXR format

### PR DESCRIPTION
**Pull request by hallonterror**
There are compression modes other than the default that are more suited for certain data. Mainly grainy/noisy data.

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [X] I agree to contribute to the project under Apache 2 License.
- [X] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [X] The PR is proposed to proper branch
- [-] There is reference to original bug report and related work
- [X] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [X] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by asmorkalov** [comment]
@hallonterror Thanks for the useful option. Couple of proposals for the solution:

File: modules/imgcodecs/src/grfmt_exr.cpp [hunk 1]
```diff
@@ -592,6 +592,44 @@ bool  ExrEncoder::write( const Mat& img, const std::vector<int>& params )
...
+            header.compression() = B44A_COMPRESSION;
+            break;
+          case IMWRITE_EXR_COMPRESSION_DWAA:
+            header.compression() = DWAA_COMPRESSION;
+            break;
+          case IMWRITE_EXR_COMPRESSION_DWAB:
+            header.compression() = DWAB_COMPRESSION;
+            break;
+          default:
+            throw std::runtime_error("IMWRITE_EXR_COMPRESSION is invalid or not supported");
```
[the remark this answers is not in the document]
Comment (hallonterror) [reply modules/imgcodecs/src/grfmt_exr.cpp:630]: Should this be done although the CV_IMWRITE_EXR_TYPE throws?

**Comment by hallonterror** [comment]
> @hallonterror Thanks for the useful option. Couple of proposals for the solution:
> 
> * OpenCV team stores test data in extra repository: https://github.com/opencv/opencv_extra/. You can add your test image there, or, better re-use one of the images, if it's applicable. So you do not need `#ifndef in test code`. If you decided to add new image to opencv_extra please submit PR there and add `opencv_extra=<your branch_name>` to the PR rescription body to apply it on BuildBot automatically.
> * The test just checks that the image is loaded correctly, but not option itself. As soon as compression option introduces significant size change I propose to add size check as simple heuristic. Please do not forget to add comment there to cover the heuristic purpose.

I know of the test data in opencv_extra. I think it should be sufficient to use one of the already existing image files (like it is in the added unit test (modules/imgcodecs/test/test_exr.impl.hpp), even though it wont show a compression advantage on those files.

For the second point you are absolutely right. I did verify that the parameter had effect exactly this way, by checking that the file size was not the same as the previously existing unit test without the flag set. It's of course reasonable to add this as a check in the test suite. I will look into that!

File: modules/imgcodecs/test/test_exr.impl.hpp [hunk 2]
```diff
@@ -113,5 +125,35 @@ TEST(Imgcodecs_EXR, readWrite_32FC3_half)
...
+  params.push_back(IMWRITE_EXR_COMPRESSION_PIZ);
+
+#ifndef GENERATE_DATA
+  const Mat img = cv::imread(filenameInput, IMREAD_UNCHANGED);
+#else
+  const Size sz(64, 32);
+  Mat img(sz, CV_32FC1, Scalar(0.5, 0.1, 1));
+  img(Rect(10, 5, sz.width - 30, sz.height - 20)).setTo(Scalar(1, 0, 0));
+  ASSERT_TRUE(cv::imwrite(filenameInput, img, params));
+#endif
```
Comment (asmorkalov) [inline modules/imgcodecs/test/test_exr.impl.hpp:145]: The else branch is not needed any more as soon as test reads test image and checks output file size. Else branch should not pass test.

File: modules/imgcodecs/test/test_exr.impl.hpp [hunk 3]
```diff
@@ -23,6 +34,7 @@ TEST(Imgcodecs_EXR, readWrite_32FC1)
     ASSERT_EQ(CV_32FC1,img.type());
 
     ASSERT_TRUE(cv::imwrite(filenameOutput, img));
+    ASSERT_EQ(getFileSize(filenameOutput), 396);
```
Comment (asmorkalov) [inline modules/imgcodecs/test/test_exr.impl.hpp:37]: Please add comment why you check file size here. It's important for test, but not obvious for other contributors.

File: modules/imgcodecs/test/test_exr.impl.hpp [hunk 4]
```diff
@@ -113,5 +125,35 @@ TEST(Imgcodecs_EXR, readWrite_32FC3_half)
...
+  const Size sz(64, 32);
+  Mat img(sz, CV_32FC1, Scalar(0.5, 0.1, 1));
+  img(Rect(10, 5, sz.width - 30, sz.height - 20)).setTo(Scalar(1, 0, 0));
+  ASSERT_TRUE(cv::imwrite(filenameInput, img, params));
+#endif
+  ASSERT_FALSE(img.empty());
+  ASSERT_EQ(CV_32FC1, img.type());
+
+  ASSERT_TRUE(cv::imwrite(filenameOutput, img, params));
+  ASSERT_EQ(getFileSize(filenameOutput), 849);
```
Comment (asmorkalov) [inline modules/imgcodecs/test/test_exr.impl.hpp:150]: The same for file size here.

File: modules/imgcodecs/src/grfmt_exr.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (asmorkalov) [reply modules/imgcodecs/src/grfmt_exr.cpp:630]: Yes.

**Review by asmorkalov — CHANGES_REQUESTED** [review]

File: modules/imgcodecs/test/test_exr.impl.hpp (hunk 2 above)
Comment (alalek) [reply modules/imgcodecs/test/test_exr.impl.hpp:145]: "Else" branch is for initial test data generation.

This code is useful for test data update in the future.

**Comment by asmorkalov** [comment]
@hallonterror Please take a look on CI status. BuildBot reports build issue:

**Comment by asmorkalov** [comment]
@hallonterror Friendly reminder.

**Comment by asmorkalov** [comment]
@hallonterror Do you have a chance to finish the patch?

**Comment by asmorkalov** [comment]
@hallonterror Friendly reminder.

**Comment by asmorkalov** [comment]
@hallonterror Unfortunatelly I cannot push changes to you  PR branch as you do it in master. I cherry picked the commits and created new PR with code review fixes: https://github.com/opencv/opencv/pull/19540

**Pull request closed without merging**